### PR TITLE
Harden against Postgres interactions with C++ destructors

### PR DESCRIPTION
**Pull request by Y--**
Fixes https://github.com/duckdb/pg_duckdb/issues/93

* make sure that no function uses `elog(ERROR...` which would bypass C++ dtors, and instead throw regular exceptions
* wraps every PG hooks in a try/catch block that converts the C++ exception to an PG equivalent (`elog(ERROR, ...` form)

### CODE REVIEW

File: src/pgduckdb_duckdb.cpp [hunk 1]
```diff
@@ -186,25 +183,20 @@ DuckDBManager::LoadExtensions(duckdb::ClientContext &context) {
...
 
 	for (auto &extension : duckdb_extensions) {
-		StringInfo duckdb_extension = makeStringInfo();
 		if (extension.enabled) {
-			appendStringInfo(duckdb_extension, "LOAD %s;", extension.name.c_str());
-			auto res = context.Query(duckdb_extension->data, false);
-			if (res->HasError()) {
-				elog(ERROR, "(PGDuckDB/LoadExtensions) `%s` could not be loaded with DuckDB", extension.name.c_str());
-			}
+			DuckDBQueryOrThrow(context, "LOAD " + extension.name);
```
Comment (hlinnaka) [inline src/pgduckdb_duckdb.cpp:187]: I think DuckDBManager::DropSecrets needs a similar change

**Review by hlinnaka — COMMENTED** [review]
At quick glance, there are calls to  get_namespace_oid() and get_relname_relid() in DuckDBManager::CheckSecretsSeq() that also seem unsafe.

Is there a general plan for how to sanity check that each function is handling errors correctly. If I understand correctly, some code runs in "Postgres regime", where you use elog(ERROR), memory context etc, while other code runs in "duckdb regime" where you use c++ exceptions etc. And whenever you cross the boundary, you must use PostgresFunctionGuard or DuckDBFunctionGuard.

Can we have some kind of compiler support or conventions or something to help with getting that right?

**Comment by hlinnaka** [comment]
> My goal is to get to a place where we:
> 
>  * use C++ as much as possible in this codebase
> 
>  * wrap non-trivial PG function calls with `PostgresFunctionGuard` to catch PG errors and throw C++ exceptions
> 
>  * wrap every top-level function (hooks, etc.) to catch C++ exceptions and convert back to PG errors
> 
> What do you think about this approach?

Yep, that works. Greenplum's orca optimizer does something similar, see e.g. https://github.com/cloudberrydb/cloudberrydb/blob/main/src/backend/gpopt/gpdbwrappers.cpp. I think PostGIS too.

**Comment by JelteF** [comment]
Moved this to the 0.2.0 milestone, to allow us to focus on stability testing for 0.1.0. We might still merge part of this PR for 0.1.0 where necessary to resolve stability issues we find during testing, but for now merging this wholesale so close to the release seems more risky than postponing it. We'll try to merge this (and related fixes) early in the 0.2.0 release cycle so it can bake longer.

File: src/pgduckdb_ddl.cpp [hunk 2]
```diff
@@ -187,18 +190,19 @@ duckdb_create_table_trigger(PG_FUNCTION_ARGS) {
...
 	Datum relid_datum = SPI_getbinval(tuple, SPI_tuptable->tupdesc, 1, &isnull);
 	if (isnull) {
-		elog(ERROR, "Expected relid to be returned, but found NULL");
+		throw std::runtime_error("Expected relid to be returned, but found NULL");
 	}
+
 	Datum is_temporary_datum = SPI_getbinval(tuple, SPI_tuptable->tupdesc, 2, &isnull);
 	if (isnull) {
-		elog(ERROR, "Expected temporary boolean to be returned, but found NULL");
+		throw std::runtime_error("Expected temporary boolean to be returned, but found NULL");
```
Comment (JelteF) [inline src/pgduckdb_ddl.cpp:205]: I think we discussed this in person, but I'll write it down here as well:

I'm not entirely convinced yet that changing all the elogs to exceptions in this file and in `pgduckdb_ruleutils.cpp` is all that helpful. It doesn't hurt either functionally, because you've wrapped this fuction in an `InvokeCPPFunc` call, and there's no `PG_TRY` around them. So there is no problem with throwing exceptions now.

But I think these specific functions probably won't benefit from being rewritten to C++ completely, and would be easier to maintain in Postgres C style. Mainly because they call a lot of Postgres functions and very few DuckDB functions. Wrapping all of those Postgres calls in `PostgresFunctionGuard` calls doesn't necessarily sound like it makes the code easier to read and maintain. I'm happily proven wrong though. 

If you're fairly sure this is the right direction for these functions I'm fine with merging this now. But I think it should be a separate commit from most of the other changes (specifically the new and improved macros). To make reverting it easier if it turns out we want to make these functions fully Postgres C instead of fully C++.

File: src/pgduckdb_table_am.cpp [hunk 3]
```diff
@@ -398,70 +399,70 @@ duckdb_scan_sample_next_tuple(TableScanDesc scan, SampleScanState *scanstate, Tu
...
-                                              .tuple_complete_speculative = duckdb_tuple_complete_speculative,
-                                              .multi_insert = duckdb_multi_insert,
-                                              .tuple_delete = duckdb_tuple_delete,
-                                              .tuple_update = duckdb_tuple_update,
-                                              .tuple_lock = duckdb_tuple_lock,
-                                              .finish_bulk_insert = duckdb_finish_bulk_insert,
+static const TableAmRoutine duckdb_methods = {
+    .type = T_TableAmRoutine,
+
+    .slot_callbacks = WRAP_CPP_FUNC(duckdb_slot_callbacks),
```
Comment (JelteF) [inline src/pgduckdb_table_am.cpp:405]: Why are these needed?

File: src/pgduckdb_ddl.cpp [hunk 4]
```diff
@@ -561,8 +558,9 @@ duckdb_alter_table_trigger(PG_FUNCTION_ARGS) {
 	SetUserIdAndSecContext(saved_userid, sec_context);
 	AtEOXact_GUC(false, save_nestlevel);
 
-	if (ret != SPI_OK_SELECT)
-		elog(ERROR, "SPI_exec failed: error code %s", SPI_result_code_string(ret));
+	if (ret != SPI_OK_SELECT) {
```
[the remark this answers is not in the document]
Comment (Y--) [reply src/pgduckdb_ddl.cpp:561]: Hmm don't we want to reset the user id etc (L558-559) before throwing? I thought about adding a version with a callback to execute before throwing but that felt overkill.

File: include/pgduckdb/pgduckdb_utils.hpp [hunk 5]
```diff
@@ -25,63 +27,59 @@ TokenizeString(char *str, const char delimiter) {
...
+	} // PG_END_TRY();
+
+	auto message = duckdb::StringUtil::Format("(PGDuckDB/%s) %s", func_name, edata->message);
+	throw duckdb::Exception(duckdb::ExceptionType::EXECUTOR, message);
 }
 
-template <typename FuncRetT, typename FuncType, typename... FuncArgs>
-FuncRetT
-DuckDBFunctionGuard(FuncType duckdb_function, const char* function_name, FuncArgs... args) {
+#define PostgresFunctionGuard(FUNC, ...)                                                                               \
```
[the remark this answers is not in the document]
Comment (Y--) [reply include/pgduckdb/pgduckdb_utils.hpp:74]: > I think it makes sense to move them the macros to a separate PR.

You mean landing a PR with just the macro definition? (basically this file's diff?)

File: src/pgduckdb_table_am.cpp (hunk 3 above)
Comment (Y--) [reply src/pgduckdb_table_am.cpp:405]: Ah I need to revert this, it was an experiment to wrap simply the table am callbacks. But since none of them have CPP code it's not necessary.

**Comment by JelteF** [comment]
Closing this in favor of all the smaller PRs that we've been doing to address this problem step-by-step

**Comment by Y--** [comment]
> Closing this in favor of all the smaller PRs that we've been doing to address this problem step-by-step

Yeah I was going to re-use this as the final PR but it doesn't really matter :-)
